[MI 4.5.0] IOException: Illegal character when forwarding multipart/related response with non-ASCII attachment MIME headers

### Description

When WSO2 MI 4.5.0 receives a multipart/related backend response containing SOAP with Attachments (SwA), and one or more attachment MIME headers (for example, Content-Disposition or Content-Location) contain raw non-ASCII characters in their values , such as accented characters like à, é, or ç , MI throws the following error and fails to forward the response to the client:


```
[2026-05-08 12:45:21,854] ERROR {PassThroughHttpSender} - Failed to submit the responseorg.apache.axis2.AxisFault: Illegal character 'ￃ'
	at org.apache.axis2.AxisFault.makeFault(AxisFault.java:430)
	at org.apache.axis2.transport.http.SOAPMessageFormatter.writeSwAMessage(SOAPMessageFormatter.java:277)
	at org.apache.axis2.transport.http.SOAPMessageFormatter.writeTo(SOAPMessageFormatter.java:77)
	at org.apache.synapse.transport.passthru.PassThroughHttpSender.submitResponse(PassThroughHttpSender.java:600)
	at org.apache.synapse.transport.passthru.PassThroughHttpSender.invoke(PassThroughHttpSender.java:293)
	at org.apache.axis2.engine.AxisEngine.send(AxisEngine.java:442)
	at org.apache.synapse.core.axis2.Axis2Sender.sendBack(Axis2Sender.java:236)
	at org.apache.synapse.mediators.builtin.RespondMediator.mediate(RespondMediator.java:46)
	at org.apache.synapse.mediators.AbstractListMediator.mediate(AbstractListMediator.java:132)
	at org.apache.synapse.mediators.AbstractListMediator.mediate(AbstractListMediator.java:79)
	at org.apache.synapse.mediators.base.SequenceMediator.mediate(SequenceMediator.java:158)
	at org.apache.synapse.core.axis2.ProxyServiceMessageReceiver.receive(ProxyServiceMessageReceiver.java:248)
	at org.apache.axis2.engine.AxisEngine.receive(AxisEngine.java:180)
	at org.apache.synapse.transport.passthru.ServerWorker.processEntityEnclosingRequest(ServerWorker.java:492)
	at org.apache.synapse.transport.passthru.ServerWorker.run(ServerWorker.java:208)
	at org.apache.axis2.transport.base.threads.NativeWorkerPool$1.run(NativeWorkerPool.java:172)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.IOException: Illegal character 'ￃ'
	at org.apache.axiom.mime.impl.axiom.MultipartWriterImpl.writeAscii(MultipartWriterImpl.java:73)
	at org.apache.axiom.mime.impl.axiom.MultipartWriterImpl.write(MultipartWriterImpl.java:129)
	at org.apache.axiom.mime.impl.axiom.MultipartWriterImpl.writePart(MultipartWriterImpl.java:89)
	at org.apache.axiom.mime.impl.axiom.MultipartWriterImpl.writePart(MultipartWriterImpl.java:144)
	at org.apache.axiom.om.impl.OMMultipartWriter.writePart(OMMultipartWriter.java:158)
	at org.apache.axis2.transport.http.SOAPMessageFormatter.writeSwAMessage(SOAPMessageFormatter.java:265)
	... 17 more

[2026-05-08 12:45:21,854] ERROR {Axis2Sender} - {proxy:MultipartReproProxy} Accept:*/*,activityid:62e079ce-7a9a-41c1-ac4f-af12d8669100,Content-Type:multipart/related; boundary="BOUNDARY"; type="text/xml"; start="<soap-part@test.example>",Host:localhost:8290,messageType:multipart/related,SOAPAction:"", Unexpected error sending message backorg.apache.axis2.AxisFault: Failed to submit the response
	at org.apache.synapse.transport.passthru.PassThroughHttpSender.handleException(PassThroughHttpSender.java:694)
	at org.apache.synapse.transport.passthru.PassThroughHttpSender.invoke(PassThroughHttpSender.java:295)
	at org.apache.axis2.engine.AxisEngine.send(AxisEngine.java:442)
	at org.apache.synapse.core.axis2.Axis2Sender.sendBack(Axis2Sender.java:236)
	at org.apache.synapse.mediators.builtin.RespondMediator.mediate(RespondMediator.java:46)
	at org.apache.synapse.mediators.AbstractListMediator.mediate(AbstractListMediator.java:132)
	at org.apache.synapse.mediators.AbstractListMediator.mediate(AbstractListMediator.java:79)
	at org.apache.synapse.mediators.base.SequenceMediator.mediate(SequenceMediator.java:158)
	at org.apache.synapse.core.axis2.ProxyServiceMessageReceiver.receive(ProxyServiceMessageReceiver.java:248)
	at org.apache.axis2.engine.AxisEngine.receive(AxisEngine.java:180)
	at org.apache.synapse.transport.passthru.ServerWorker.processEntityEnclosingRequest(ServerWorker.java:492)
	at org.apache.synapse.transport.passthru.ServerWorker.run(ServerWorker.java:208)
	at org.apache.axis2.transport.base.threads.NativeWorkerPool$1.run(NativeWorkerPool.java:172)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.apache.axis2.AxisFault: Illegal character 'ￃ'
	at org.apache.axis2.AxisFault.makeFault(AxisFault.java:430)
	at org.apache.axis2.transport.http.SOAPMessageFormatter.writeSwAMessage(SOAPMessageFormatter.java:277)
	at org.apache.axis2.transport.http.SOAPMessageFormatter.writeTo(SOAPMessageFormatter.java:77)
	at org.apache.synapse.transport.passthru.PassThroughHttpSender.submitResponse(PassThroughHttpSender.java:600)
	at org.apache.synapse.transport.passthru.PassThroughHttpSender.invoke(PassThroughHttpSender.java:293)
	... 14 more
Caused by: java.io.IOException: Illegal character 'ￃ'
	at org.apache.axiom.mime.impl.axiom.MultipartWriterImpl.writeAscii(MultipartWriterImpl.java:73)
	at org.apache.axiom.mime.impl.axiom.MultipartWriterImpl.write(MultipartWriterImpl.java:129)
	at org.apache.axiom.mime.impl.axiom.MultipartWriterImpl.writePart(MultipartWriterImpl.java:89)
	at org.apache.axiom.mime.impl.axiom.MultipartWriterImpl.writePart(MultipartWriterImpl.java:144)
	at org.apache.axiom.om.impl.OMMultipartWriter.writePart(OMMultipartWriter.java:158)
	at org.apache.axis2.transport.http.SOAPMessageFormatter.writeSwAMessage(SOAPMessageFormatter.java:265)
	... 17 more

```


### Steps to Reproduce

1.) Deploy the following proxy in MI 4.5.0:

```
<proxy name="MultipartReproProxy"
       xmlns="http://ws.apache.org/ns/synapse"
       transports="http https"
       startOnLoad="true">
    <target>
        <inSequence>
            <log category="INFO" level="full"/>  
            <property name="enableSwA" scope="axis2" value="true"/>
            <respond/>
        </inSequence>
    </target>
</proxy>
```

2.) Invoke using the following curl command:

```
curl -s -X POST http://localhost:8290/services/MultipartReproProxy \
  -H 'Content-Type: multipart/related; boundary="BOUNDARY"; type="text/xml"; start="<soap-part@test.example>"' \
  -H 'SOAPAction: ""' \
  --data-binary @/tmp/swa_request.bin
```

3.) The swa_request.bin file content is provided below.

```
--BOUNDARY
Content-Type: text/xml; charset=UTF-8
Content-ID: <soap-part@test.example>

<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
    <soapenv:Body>
        <test/>
    </soapenv:Body>
</soapenv:Envelope>

--BOUNDARY
Content-Type: application/pdf
Content-Transfer-Encoding: binary
Content-ID: 123456
Content-Disposition: attachment; filename="invoàààice.pdf"

%PDF-1.4 fake

--BOUNDARY--
```


### Version

Mi 4.5.0

### Environment Details (with versions)

_No response_

--- a/modules/kernel/src/org/apache/axis2/transport/http/SOAPMessageFormatter.java
+++ b/modules/kernel/src/org/apache/axis2/transport/http/SOAPMessageFormatter.java
@@ -26,7 +26,10 @@ import org.apache.axiom.om.impl.OMMultipartWriter;
 import org.apache.axiom.util.UIDGenerator;
 import org.apache.axis2.AxisFault;
 import org.apache.axis2.Constants;
+import org.apache.axis2.context.ConfigurationContext;
 import org.apache.axis2.context.MessageContext;
+import org.apache.axis2.description.Parameter;
+import org.apache.axis2.engine.AxisConfiguration;
 import org.apache.axis2.transport.MessageFormatter;
 import org.apache.axis2.transport.http.util.URLTemplatingUtil;
 import org.apache.axis2.util.JavaUtils;
@@ -49,6 +52,7 @@ public class SOAPMessageFormatter implements MessageFormatter {
 
     private static final Log log = LogFactory.getLog(SOAPMessageFormatter.class);
     private static final String WRITE_XML_DECLARATION = "WRITE_XML_DECLARATION";
+    private static final String DROP_ATTACHMENT_HEADERS = "drop_attachment_headers";
     
     public void writeTo(MessageContext msgCtxt, OMOutputFormat format,
                         OutputStream out, boolean preserve) throws AxisFault {
@@ -245,17 +249,51 @@ public class SOAPMessageFormatter implements MessageFormatter {
             
             Attachments attachments = msgCtxt.getAttachmentMap();
             Map headersMap = attachments.getHeaders();
+
+            // Server wide opt in (axis2.xml <parameter name="drop_attachment_headers">true</parameter>):
+            // restore pre EI-5440 / ESB 4.9.0 behavior by skipping all attachment MIME headers.
+            // Default false — headers are still preserved per EI-5440, but sanitized to drop any
+            // single header whose name/value contains non-ASCII chars that would trip
+            // axiom's MultipartWriterImpl.writeAscii(). Read directly from AxisConfiguration
+            // because msgCtxt.getParameter() may not resolve transport level parameters reliably.
+            boolean dropAllHeaders = false;
+            ConfigurationContext configurationContext = msgCtxt.getConfigurationContext();
+            if (configurationContext != null) {
+                AxisConfiguration axisConfig = configurationContext.getAxisConfiguration();
+                if (axisConfig != null) {
+                    Parameter dropHeadersParam = axisConfig.getParameter(DROP_ATTACHMENT_HEADERS);
+                    if (dropHeadersParam != null) {
+                        dropAllHeaders = JavaUtils.isTrueExplicitly(dropHeadersParam.getValue());
+                    }
+                }
+            }
+
             for (String contentID : attachments.getAllContentIDs()) {
                 if (!contentID.equalsIgnoreCase(attachments.getSOAPPartContentID())) {
                     List<String> headersList = new ArrayList<>();
-                    Object o = headersMap.get(contentID);
-                    if (o instanceof List) {
-                        List currentHeaders = (List) o;
-                        for (Object currentHeader : currentHeaders) {
-                            if (currentHeader instanceof Header) {
+                    if (!dropAllHeaders) {
+                        Object o = headersMap.get(contentID);
+                        if (o instanceof List) {
+                            for (Object currentHeader : (List) o) {
+                                if (!(currentHeader instanceof Header)) {
+                                    continue;
+                                }
                                 Header header = (Header) currentHeader;
-                                headersList.add(header.getName() + ": " + header.getValue());
-
+                                String name = header.getName();
+                                String value = header.getValue();
+                                // Axiom's MultipartWriterImpl emits Content-Type,
+                                // Content-Transfer-Encoding and Content-ID itself. Forwarding
+                                // them again would produce duplicate MIME headers (RFC 2045 §3).
+                                if (isStructuralMimeHeader(name)) {
+                                    continue;
+                                }
+                                if (isAscii(name) && isAscii(value)) {
+                                    headersList.add(name + ": " + value);
+                                } else if (log.isWarnEnabled()) {
+                                    log.warn("Dropping non-ASCII MIME header '" + name
+                                            + "' on attachment " + contentID
+                                            + "; header names and values must be ASCII per RFC 2045.");
+                                }
                             }
                         }
                     }
@@ -282,6 +320,42 @@ public class SOAPMessageFormatter implements MessageFormatter {
         if (log.isDebugEnabled()) {
             log.debug("end writeSwAMessage()");
         }
+    }
+
+    /**
+     * Checks whether the header is a structural MIME header emitted by the multipart writer.
+     *
+     * @param name the MIME header name to check
+     * @return {@code true} if the header is managed by the multipart writer
+     */
+    private static boolean isStructuralMimeHeader(String name) {
+        return "Content-Type".equalsIgnoreCase(name)
+                || "Content-Transfer-Encoding".equalsIgnoreCase(name)
+                || "Content-ID".equalsIgnoreCase(name);
+    }
+
+    /**
+     * Checks if a string contains only ASCII characters (code points 0-127).
+     * <p>
+     * This method is used to validate MIME header names and values before writing them
+     * to ensure they don't contain non-ASCII characters that would cause issues with
+     * the multipart writer implementation.
+     * </p>
+     *
+     * @param headerValue the string to check; may be null
+     * @return {@code true} if the string is null or contains only ASCII characters;
+     *         {@code false} if any character has a code point >= 128
+     */
+    private static boolean isAscii(String headerValue) {
+        if (headerValue == null) {
+            return true;
+        }
+        for (int i = 0, len = headerValue.length(); i < len; i++) {
+            if (headerValue.charAt(i) >= 128) {
+                return false;
+            }
+        }
+        return true;
     }
 
 }

